File: src/cpp/surface_mesh.cpp
#include <pybind11/eigen.h>
#include <pybind11/numpy.h>
#include <pybind11/pybind11.h>

#include "Eigen/Dense"

#include "polyscope/polyscope.h"
#include "polyscope/surface_mesh.h"

namespace py = pybind11;
namespace ps = polyscope;


void bind_surface_mesh(py::module& m) {
  
  
  // == Helper quantity classes
  
  // Scalar quantities 
  py::class_<ps::SurfaceVertexScalarQuantity>(m, "SurfaceVertexScalarQuantity")
    .def("set_enabled", &ps::SurfaceVertexScalarQuantity::setEnabled, "Set enabled")
    .def("set_color_map", &ps::SurfaceVertexScalarQuantity::setColorMap, "Set color map")
    .def("set_map_range", &ps::SurfaceVertexScalarQuantity::setMapRange, "Set map range");
  py::class_<ps::SurfaceFaceScalarQuantity>(m, "SurfaceFaceScalarQuantity")
    .def("set_enabled", &ps::SurfaceFaceScalarQuantity::setEnabled, "Set enabled")
    .def("set_color_map", &ps::SurfaceFaceScalarQuantity::setColorMap, "Set color map")
    .def("set_map_range", &ps::SurfaceFaceScalarQuantity::setMapRange, "Set map range");
  py::class_<ps::SurfaceEdgeScalarQuantity>(m, "SurfaceEdgeScalarQuantity")
    .def("set_enabled", &ps::SurfaceEdgeScalarQuantity::setEnabled, "Set enabled")
    .def("set_color_map", &ps::SurfaceEdgeScalarQuantity::setColorMap, "Set color map")
    .def("set_map_range", &ps::SurfaceEdgeScalarQuantity::setMapRange, "Set map range");
  py::class_<ps::SurfaceHalfedgeScalarQuantity>(m, "SurfaceHalfedgeScalarQuantity")
    .def("set_enabled", &ps::SurfaceHalfedgeScalarQuantity::setEnabled, "Set enabled")
    .def("set_color_map", &ps::SurfaceHalfedgeScalarQuantity::setColorMap, "Set color map")
    .def("set_map_range", &ps::SurfaceHalfedgeScalarQuantity::setMapRange, "Set map range");
  
  // Color quantities 
  py::class_<ps::SurfaceVertexColorQuantity>(m, "SurfaceVertexColorQuantity")
    .def("set_enabled", &ps::SurfaceVertexColorQuantity::setEnabled, "Set enabled");
  py::class_<ps::SurfaceFaceColorQuantity>(m, "SurfaceFaceColorQuantity")
    .def("set_enabled", &ps::SurfaceFaceColorQuantity::setEnabled, "Set enabled");
  
  // Distance quantities 
  py::class_<ps::SurfaceDistanceQuantity>(m, "SurfaceDistanceQuantity")
    .def("set_enabled", &ps::SurfaceDistanceQuantity::setEnabled, "Set enabled")
    .def("set_color_map", &ps::SurfaceDistanceQuantity::setColorMap, "Set color map")
    .def("set_map_range", &ps::SurfaceDistanceQuantity::setMapRange, "Set map range")
    .def("set_stripe_size", &ps::SurfaceDistanceQuantity::setStripeSize, "Set stripe size");


  // == Main class
  py::class_<ps::SurfaceMesh>(m, "SurfaceMesh")
    
    // basics
    .def("remove", &ps::SurfaceMesh::remove, "Remove the structure")
    .def("set_enabled", &ps::SurfaceMesh::setEnabled, "Enable the structure")
    .def("is_enabled", &ps::SurfaceMesh::isEnabled, "Check if the structure is enabled")
    .def("remove_all_quantities", &ps::SurfaceMesh::removeAllQuantities, "Remove all quantities")
    .def("remove_quantity", &ps::SurfaceMesh::removeQuantity, "Remove a quantity")
    .def("update_vertex_positions", &ps::SurfaceMesh::updateVertexPositions<Eigen::MatrixXd>, "Update vertex positions")
    .def("update_vertex_positions2D", &ps::SurfaceMesh::updateVertexPositions2D<Eigen::MatrixXd>, "Update vertex positions")
    .def("n_vertices", &ps::SurfaceMesh::nVertices, "# vertices")
    .def("n_faces", &ps::SurfaceMesh::nFaces, "# edges")
    .def("n_edges", &ps::SurfaceMesh::nEdges, "# edges")
    .def("n_corners", &ps::SurfaceMesh::nCorners, "# corners")
    .def("n_halfedges", &ps::SurfaceMesh::nHalfedges, "# halfedges")

    // options
    .def("set_color", &ps::SurfaceMesh::setSurfaceColor, "Set surface color")
    .def("get_color", &ps::SurfaceMesh::getSurfaceColor, "Get surface color")
    .def("set_edge_color", &ps::SurfaceMesh::setEdgeColor, "Set edge color")
    .def("get_edge_color", &ps::SurfaceMesh::getEdgeColor, "Get edge color")
    .def("set_edge_width", &ps::SurfaceMesh::setEdgeWidth, "Set edge width")
    .def("get_edge_width", &ps::SurfaceMesh::getEdgeWidth, "Get edge width")
    .def("set_smooth_shade", &ps::SurfaceMesh::setSmoothShade, "Set smooth shading")
    .def("get_smooth_shade", &ps::SurfaceMesh::isSmoothShade, "Get if smooth shading is enabled")
    .def("set_material", &ps::SurfaceMesh::setMaterial, "Set material")
    .def("get_material", &ps::SurfaceMesh::getMaterial, "Get material")

    // permutations & bases
    .def("set_vertex_permutation", &ps::SurfaceMesh::setVertexPermutation<Eigen::VectorXi>, "Set vertex permutation")
    .def("set_face_permutation", &ps::SurfaceMesh::setFacePermutation<Eigen::VectorXi>, "Set face permutation")
    .def("set_edge_permutation", &ps::SurfaceMesh::setEdgePermutation<Eigen::VectorXi>, "Set edge permutation")
    .def("set_halfedge_permutation", &ps::SurfaceMesh::setHalfedgePermutation<Eigen::VectorXi>, "Set halfedge permutation")
    .def("set_corner_permutation", &ps::SurfaceMesh::setCornerPermutation<Eigen::VectorXi>, "Set corner permutation")
    .def("set_vertex_tangent_basisX", &ps::SurfaceMesh::setVertexTangentBasisX<Eigen::MatrixXd>, "Set vertex tangent bases")
    .def("set_vertex_tangent_basisX2D", &ps::SurfaceMesh::setVertexTangentBasisX2D<Eigen::MatrixXd>, "Set vertex tangent bases")
    .def("set_face_tangent_basisX", &ps::SurfaceMesh::setFaceTangentBasisX<Eigen::MatrixXd>, "Set face tangent bases")
    .def("set_face_tangent_basisX2D", &ps::SurfaceMesh::setFaceTangentBasisX2D<Eigen::MatrixXd>, "Set face tangent bases")

    // = quantities
    
    // Scalars 
    .def("add_vertex_scalar_quantity", &ps::SurfaceMesh::addVertexScalarQuantity<Eigen::VectorXd>, "Add a scalar function at vertices",
        py::arg("name"), py::arg("values"), py::arg("data_type")=ps::DataType::STANDARD, py::return_value_policy::reference)
    .def("add_face_scalar_quantity", &ps::SurfaceMesh::addFaceScalarQuantity<Eigen::VectorXd>, "Add a scalar function at faces",
        py::arg("name"), py::arg("values"), py::arg("data_type")=ps::DataType::STANDARD, py::return_value_policy::reference)
    .def("add_edge_scalar_quantity", &ps::SurfaceMesh::addEdgeScalarQuantity<Eigen::VectorXd>, "Add a scalar function at edges",
        py::arg("name"), py::arg("values"), py::arg("data_type")=ps::DataType::STANDARD, py::return_value_policy::reference)
    .def("add_halfedge_scalar_quantity", &ps::SurfaceMesh::addHalfedgeScalarQuantity<Eigen::VectorXd>, "Add a scalar function at halfedges",
        py::arg("name"), py::arg("values"), py::arg("data_type")=ps::DataType::STANDARD, py::return_value_policy::reference)
    
    // Colors
    .def("add_vertex_color_quantity", &ps::SurfaceMesh::addVertexColorQuantity<Eigen::MatrixXd>, "Add a color value at vertices",
        py::return_value_policy::reference)
    .def("add_face_color_quantity", &ps::SurfaceMesh::addFaceColorQuantity<Eigen::MatrixXd>, "Add a color value at faces",
        py::return_value_policy::reference)
    
    // Distance
    .def("add_vertex_distance_quantity", &ps::SurfaceMesh::addVertexDistanceQuantity<Eigen::VectorXd>, 
        "Add a distance function at vertices", py::return_value_policy::reference)
    .def("add_vertex_signed_distance_quantity", &ps::SurfaceMesh::addVertexSignedDistanceQuantity<Eigen::VectorXd>, 
        "Add a signed distance function at vertices", py::return_value_policy::reference)

    // Vector
    .def("add_vertex_vector_quantity", &ps::SurfaceMesh::addVertexVectorQuantity<Eigen::MatrixXd>, "Add a vector function at vertices",
        py::arg("name"), py::arg("values"), py::arg("vector_type")=ps::VectorType::STANDARD, py::return_value_policy::reference)
    .def("add_face_vector_quantity", &ps::SurfaceMesh::addFaceVectorQuantity<Eigen::MatrixXd>, "Add a vector function at faces",
        py::arg("name"), py::arg("values"), py::arg("vector_type")=ps::VectorType::STANDARD, py::return_value_policy::reference);

  
  // Static adders and getters
  m.def("register_surface_mesh", &ps::registerSurfaceMesh<Eigen::MatrixXd, Eigen::MatrixXi>, 
      py::arg("name"), py::arg("nodes"), py::arg("edges"),
      "Register a surface mesh", py::return_value_policy::reference);
  m.def("register_surface_mesh2D", &ps::registerSurfaceMesh2D<Eigen::MatrixXd, Eigen::MatrixXi>, 
      py::arg("name"), py::arg("nodes"), py::arg("edges"),
      "Register a surface mesh", py::return_value_policy::reference);

  m.def("remove_surface_mesh", &polyscope::removeSurfaceMesh, "Remove a surface mesh by name");
  m.def("get_surface_mesh", &polyscope::getSurfaceMesh, "Get a surface mesh by name", py::return_value_policy::reference);
  m.def("has_surface_mesh", &polyscope::hasSurfaceMesh, "Check for a surface mesh by name");
}

File: src/polyscope/surface_mesh.py
import polyscope_bindings as psb

from polyscope.core import str_to_datatype, str_to_vectortype, glm3

class SurfaceMesh:

    # This class wraps a _reference_ to the underlying object, whose lifetime is managed by Polyscope

    # End users should not call this constrctor, use register_surface_mesh instead
    def __init__(self, name=None, vertices=None, faces=None, instance=None):


        # TODO ragged vertex arrays

        if instance is not None:
            # Wrap an existing instance
            self.bound_mesh = instance

        else:
            # Create a new instance
            self.check_shape(vertices)

            if (len(faces.shape) != 2):
                raise ValueError("surface mesh face array should have shape (F,D) for some D; shape is " + str(faces.shape))

            if vertices.shape[1] == 3:
                self.bound_mesh = psb.register_surface_mesh(name, vertices, faces) 
            elif vertices.shape[1] == 2:
                self.bound_mesh = psb.register_surface_mesh2D(name, vertices, faces) 

    def check_shape(self, points):
        # Helper to validate arrays
        if (len(points.shape) != 2) or (points.shape[1] not in (2,3)):
            raise ValueError("surface mesh node positions should have shape (N,3) or (N,2); shape is " + str(points.shape))
      

    def n_vertices(self):
        return self.bound_mesh.n_vertices()
    def n_faces(self):
        return self.bound_mesh.n_faces()
    def n_edges(self):
        return self.bound_mesh.n_edges()
    def n_halfedges(self):
        return self.bound_mesh.n_halfedges()
    def n_corners(self):
        return self.bound_mesh.n_corners()

    ## Structure management
    
    def remove(self):
        '''Remove the structure itself'''
        self.bound_mesh.remove()
    def remove_all_quantities(self):
        '''Remove all quantities on the structure'''
        self.bound_mesh.remove_all_quantities()
    def remove_quantity(self, name):
        '''Remove a single quantity on the structure'''
        self.bound_mesh.remove_quantity(name)

    # Enable/disable
    def set_enabled(self, val=True):
        self.bound_mesh.set_enabled(val)
    def is_enabled(self):
        return self.bound_mesh.is_enabled()

    # Update
    def update_vertex_positions(self, vertices):
        self.check_shape(vertices)
        
        if vertices.shape[1] == 3:
            self.bound_mesh.update_vertex_positions(vertices)
        elif vertices.shape[1] == 2:
            self.bound_mesh.update_vertex_positions2D(vertices)
        else:
            raise ValueError("bad node shape")

    ## Options

    
    # Color
    def set_color(self, val):
        self.bound_mesh.set_color(glm3(val))
    def get_color(self):
        return self.bound_mesh.get_color().as_tuple()
    
    # Edge Color
    def set_edge_color(self, val):
        self.bound_mesh.set_edge_color(glm3(val))
    def get_edge_color(self):
        return self.bound_mesh.get_edge_color().as_tuple()
    
    # Edge width
    def set_edge_width(self, val):
        self.bound_mesh.set_edge_width(val)
    def get_edge_width(self):
        return self.bound_mesh.get_edge_width()
    
    # Smooth shade
    def set_smooth_shade(self, val):
        self.bound_mesh.set_smooth_shade(val)
    def get_smooth_shade(self):
        return self.bound_mesh.get_smooth_shade()
    
    # Material
    def set_material(self, mat):
        self.bound_mesh.set_material(mat)
    def get_material(self):
        return self.bound_mesh.get_material()


    ## Permutations and bases

    def set_vertex_permutation(self, perm, expected_size=None):
        if len(perm.shape) != 1 or perm.shape[0] != self.n_vertices(): raise ValueError("'perm' should be an array with one entry per vertex")
        if expected_size is None: expected_size = 0
        self.bound_mesh.set_vertex_permutation(perm, expected_size)

    def set_face_permutation(self, perm, expected_size=None):
        if len(perm.shape) != 1 or perm.shape[0] != self.n_faces(): raise ValueError("'perm' should be an array with one entry per face")
        if expected_size is None: expected_size = 0
        self.bound_mesh.set_face_permutation(perm, expected_size)
    
    def set_edge_permutation(self, perm, expected_size=None):
        if len(perm.shape) != 1 or perm.shape[0] != self.n_edges(): raise ValueError("'perm' should be an array with one entry per edge")
        if expected_size is None: expected_size = 0
        self.bound_mesh.set_edge_permutation(perm, expected_size)
    
    def set_corner_permutation(self, perm, expected_size=None):
        if len(perm.shape) != 1 or perm.shape[0] != self.n_corners(): raise ValueError("'perm' should be an array with one entry per corner")
        if expected_size is None: expected_size = 0
        self.bound_mesh.set_corner_permutation(perm, expected_size)
    
    def set_halfedge_permutation(self, perm, expected_size=None):
        if len(perm.shape) != 1 or perm.shape[0] != self.n_halfedges(): raise ValueError("'perm' should be an array with one entry per halfedge")
        if expected_size is None: expected_size = 0
        self.bound_mesh.set_halfedge_permutation(perm, expected_size)
    
    def set_all_permutations(self, 
            vertex_perm=None, vertex_perm_size=None,
            face_perm=None, face_perm_size=None,
            edge_perm=None, edge_perm_size=None,
            corner_perm=None, corner_perm_size=None,
            halfedge_perm=None, halfedge_perm_size=None):

        if vertex_perm is not None: self.set_vertex_permutation(vertex_perm, vertex_perm_size)
        if face_perm is not None: self.set_face_permutation(face_perm, face_perm_size)
        if edge_perm is not None: self.set_edge_permutation(edge_perm, edge_perm_size)
        if corner_perm is not None: self.set_corner_permutation(corner_perm, corner_perm_size)
        if halfedge_perm is not None: self.set_halfedge_permutation(halfedge_perm, halfedge_perm_size)
    
    def set_vertex_tangent_basisX(self, vectors):
        if len(vectors.shape) != 2 or vectors.shape[0] != self.n_vertices() or vectors.shape[1] not in (2,3): 
            raise ValueError("'vectors' should be an array with one entry per vertex")

        if vectors.shape[1] == 2:
            self.bound_mesh.set_vertex_tangent_basisX2D(vectors)
        elif vectors.shape[1] == 3:
            self.bound_mesh.set_vertex_tangent_basisX(vectors)
    
    def set_face_tangent_basisX(self, vectors):
        if len(vectors.shape) != 2 or vectors.shape[0] != self.n_faces() or vectors.shape[1] not in (2,3): 
            raise ValueError("'vectors' should be an array with one entry per face")

        if vectors.shape[1] == 2:
            self.bound_mesh.set_face_tangent_basisX2D(vectors)
        elif vectors.shape[1] == 3:
            self.bound_mesh.set_face_tangent_basisX(vectors)
    
         


    ## Quantities

    # Scalar
    def add_scalar_quantity(self, name, values, defined_on='vertices', enabled=None, datatype="standard", vminmax=None, cmap=None):

        if len(values.shape) != 1: raise ValueError("'values' should be a length-N array")

        if defined_on == 'vertices':
            if values.shape[0] != self.n_vertices(): raise ValueError("'values' should be a length n_vertices array")
            q = self.bound_mesh.add_vertex_scalar_quantity(name, values, str_to_datatype(datatype))
        elif defined_on == 'faces':
            if values.shape[0] != self.n_faces(): raise ValueError("'values' should be a length n_faces array")
            q = self.bound_mesh.add_face_scalar_quantity(name, values, str_to_datatype(datatype))
        elif defined_on == 'edges':
            if values.shape[0] != self.n_edges(): raise ValueError("'values' should be a length n_edges array")
            q = self.bound_mesh.add_edge_scalar_quantity(name, values, str_to_datatype(datatype))
        elif defined_on == 'halfedges':
            if values.shape[0] != self.n_halfedges(): raise ValueError("'values' should be a length n_halfedges array")
            q = self.bound_mesh.add_halfedge_scalar_quantity(name, values, str_to_datatype(datatype))
        else:
            raise ValueError("bad `defined_on` value {}, should be one of ['vertices', 'faces', 'edges', 'halfedges']".format(defined_on))
            

        # Support optional params
        if enabled is not None:
            q.set_enabled(enabled)
        if vminmax is not None:
            q.set_map_range(vminmax)
        if cmap is not None:
            q.set_color_map(cmap)
    
    
    # Color
    def add_color_quantity(self, name, values, defined_on='vertices', enabled=None):
        if len(values.shape) != 2 or values.shape[1] != 3: raise ValueError("'values' should be an Nx3 array")
        
        if defined_on == 'vertices':
            if values.shape[0] != self.n_vertices(): raise ValueError("'values' should be a length n_vertices array")
            q = self.bound_mesh.add_vertex_color_quantity(name, values)
        elif defined_on == 'faces':
            if values.shape[0] != self.n_faces(): raise ValueError("'values' should be a length n_faces array")
            q = self.bound_mesh.add_face_color_quantity(name, values)
        else:
            raise ValueError("bad `defined_on` value {}, should be one of ['vertices', 'faces']".format(defined_on))

        # Support optional params
        if enabled is not None:
            q.set_enabled(enabled)
    
    
    # Distance
    def add_distance_quantity(self, name, values, defined_on='vertices', enabled=None, signed=False, vminmax=None, stripe_size=None, stripe_size_relative=True, cmap=None):

        if len(values.shape) != 1: raise ValueError("'values' should be a length-N array")

        if defined_on == 'vertices':
            if values.shape[0] != self.n_vertices(): raise ValueError("'values' should be a length n_vertices array")

            if signed:
                q = self.bound_mesh.add_vertex_signed_distance_quantity(name, values)
            else:
                q = self.bound_mesh.add_vertex_distance_quantity(name, values)
        else:
            raise ValueError("bad `defined_on` value {}, should be one of ['vertices']".format(defined_on))
            

        # Support optional params
        if enabled is not None:
            q.set_enabled(enabled)
        if stripe_size is not None:
            q.set_stripe_size(stripe_size, stripe_size_relative)
        if vminmax is not None:
            q.set_map_range(vminmax)
        if cmap is not None:
            q.set_color_map(cmap)
    
    '''
    
    # Vector
    def add_vector_quantity(self, name, values, defined_on='vertices', enabled=None, vectortype="standard", length=None, radius=None, color=None):
        if len(values.shape) != 2 or values.shape[1] not in [2,3]: raise ValueError("'values' should be an Nx3 array (or Nx2 for 2D)")
        
        
        if defined_on == 'vertices':
            if values.shape[0] != self.n_nodes(): raise ValueError("'values' should be a length n_nodes array")

            if values.shape[1] == 2:
                q = self.bound_mesh.add_node_vector_quantity2D(name, values, str_to_vectortype(vectortype))
            elif values.shape[1] == 3:
                q = self.bound_mesh.add_node_vector_quantity(name, values, str_to_vectortype(vectortype))

        elif defined_on == 'faces':
            if values.shape[0] != self.n_edges(): raise ValueError("'values' should be a length n_edges array")
            
            if values.shape[1] == 2:
                q = self.bound_mesh.add_edge_vector_quantity2D(name, values, str_to_vectortype(vectortype))
            elif values.shape[1] == 3:
                q = self.bound_mesh.add_edge_vector_quantity(name, values, str_to_vectortype(vectortype))

        else:
            raise ValueError("bad `defined_on` value {}, should be one of ['vertices', 'faces']".format(defined_on))

        # Support optional params
        if enabled is not None:
            q.set_enabled(enabled)
        if length is not None:
            q.set_length(length, True)
        if radius is not None:
            q.set_radius(radius, True)
        if color is not None:
            q.set_color(glm3(color))
    '''


def register_surface_mesh(name, vertices, faces, enabled=None, color=None, edge_color=None, smooth_shade=None, 
                          edge_width=None, material=None):
    """Register a new surface mesh"""

    p = SurfaceMesh(name, vertices, faces)

    # == Apply options
    if enabled is not None:
        p.set_enabled(enabled)
    if color is not None:
        p.set_color(color)
    if edge_color is not None:
        p.set_edge_color(edge_color)
    if edge_width is not None:
        p.set_edge_width(edge_width)
    if smooth_shade is not None:
        p.set_smooth_shade(smooth_shade)
    if material is not None:
        p.set_material(material)

    return p

def remove_surface_mesh(name, error_if_absent=True):
    """Remove a surface mesh by name"""
    psb.remove_surface_mesh(name, error_if_absent)

def get_surface_mesh(name):
    """Get surface mesh by name"""
    if not has_surface_mesh(name):
        raise ValueError("no surface mesh with name " + str(name))

    raw = psb.get_surface_mesh(name)

    # Wrap the instance
    return SurfaceMesh(instance=raw)

def has_surface_mesh(name):
    """Check if a surface mesh exists by name"""
    return psb.has_surface_mesh(name)
    